### PR TITLE
#10281: Use Cache Options functionality extended to be applied on all the layers from a service

### PR DESCRIPTION
**Pull request by mahmoudadel54**
## Description
In this PR, the functionality for use cache options for mapviewer, dashboard, geostory and context are added. Now user can specify the 'use cache options' from the catalog instead of manually setting in layer settings to activate the tile grid props.

- In includes a checkbox in catalogs for the WMS tiled services for 'use cache options' flag.
- any layer will be added from catalog with type wms service, will contain the tile gird props if 'use cache options' flag = true. 

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue
#10281

**What is the current behavior?**
#10281

**What is the new behavior?**
Now user can specify the 'use cache options' from the catalog instead of manually setting in layer settings.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/plugins/widgetbuilder/MapBuilder.jsx [hunk 1]
```diff
@@ -51,8 +55,23 @@ const chooseMapEnhancer = compose(
                 manageLayers,
                 withHandlers({
                     onLayerChoice: ({ toggleLayerSelector = () => { }, addLayer = () => { } }) => (layer) => {
-                        addLayer(layer);
-                        toggleLayerSelector(false);
+                        // fetching 'tileGridData' if layer has truthy flag 'remoteTileGrids' and adding the required props to layer object
+                        let tileGridPromise = layer.type === 'wms' && layer.remoteTileGrids ? getLayerTileMatrixSetsInfo(generateGeoServerWMTSUrl(layer), layer.name, layer) : new Promise((resolve) => resolve(null));
```
Comment (MV88) [inline web/client/plugins/widgetbuilder/MapBuilder.jsx:59]: looks nice this part

**Review by MV88 — APPROVED** [review]
@mahmoudadel54 looks good to me, nice job

**Comment by MV88** [comment]
@ElenaGallo please test it in DEV

**Comment by ElenaGallo** [comment]
@mahmoudadel54 if the custom tile grids are not available for a layer a note does not appear, it is visible only when I click on *Update available tile grids*.


https://github.com/geosolutions-it/MapStore2/assets/56537133/1bcd3d20-68be-4624-98c9-184e764f61ab

**Comment by tdipisa** [comment]
@mahmoudadel54 the behavior above reports an inconsistency. In the _display tab_ the correct state should be reported to the user and the layer configured accordingly. If _Use remote custom tile grids_ in catalog is active, I guess you have to perform a WMTS request before adding the layer to the map to verify if the remote custom tile grid definition exists and can be used. Then, the state of the layer should be consistent with this as well as the UI state.

**Comment by ElenaGallo** [comment]
Test passed, @mahmoudadel54 please backport to 2024.01.xx. Thanks

**Pull request merged**